File: leetcode-java/src/classify/search/SurroundedRegions.java
package classify.search;

public class SurroundedRegions {
    private int row;
    private int col;

    /**
     * 211ms 5.03%
     * 47.5MB 8.15%
     */
    public void solve(char[][] board) {
        if (board == null) return;
        row = board.length;
        if (row == 0) return;
        col = board[0].length;
        if (col == 0) return;

        boolean[][] visited;

        for (int i = 1; i < row - 1; i++) {
            for (int j = 1; j < col - 1; j++) {
                visited = new boolean[row][col];
                if (board[i][j] == 'O' && !visited[i][j] && !touchBorder(board, visited, i, j)) {
                    convert(board, i, j);
                }
            }
        }
    }

    private boolean touchBorder(char[][] board, boolean[][] visited, int i, int j) {
        if (i < 0 || j < 0 || i >= row || j >= col) {
            return true;
        }

        if (visited[i][j] || board[i][j] == 'X') {
            return false;
        }

        visited[i][j] = true;

        return touchBorder(board, visited, i + 1, j)
                || touchBorder(board, visited, i - 1, j)
                || touchBorder(board, visited, i, j + 1)
                || touchBorder(board, visited, i, j - 1);
    }

    private void convert(char[][] board, int i, int j) {
        if (i < 0 || j < 0 || i >= row || j >= col || board[i][j] != 'O') {
            return;
        }

        board[i][j] = 'X';
        convert(board, i + 1, j);
        convert(board, i - 1, j);
        convert(board, i, j + 1);
        convert(board, i, j - 1);
    }


    /**
     * 1ms 100.00%
     * 40.2MB 91.77%
     */
    public void solve2(char[][] board) {
        if (board.length < 2 || board[0].length < 2)
            return;

        int row = board.length;
        int col = board[0].length;

        for (int i = 0; i < row; i++) {
            if (board[i][0] == 'O') {
                boundaryDFS(board, i, 0);
            }
            if (board[i][col - 1] == 'O') {
                boundaryDFS(board, i, col - 1);
            }
        }

        for (int j = 0; j < col; j++) {
            if (board[0][j] == 'O') {
                boundaryDFS(board, 0, j);
            }
            if (board[row - 1][j] == 'O') {
                boundaryDFS(board, row - 1, j);
            }
        }

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                } else if (board[i][j] == '*') {
                    board[i][j] = 'O';
                }
            }
        }
    }

    private void boundaryDFS(char[][] board, int i, int j) {
        if (i < 0 || i > board.length - 1 || j < 0 || j > board[0].length - 1 || board[i][j] != 'O') {
            return;
        }

        board[i][j] = '*';
        boundaryDFS(board, i + 1, j);
        boundaryDFS(board, i - 1, j);
        boundaryDFS(board, i, j + 1);
        boundaryDFS(board, i, j - 1);
    }
}

File: leetcode-scala/src/dfs/medium/SurroundedRegions.scala
package dfs.medium

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/7/20 22:33
 * @Description SurroundedRegions
 */
object SurroundedRegions {

    /**
     * 656ms    100.00%
     * 56.1MB   100.00%
     * dfs递归
     * @param board
     */
    def solve(board: Array[Array[Char]]): Unit = {
        if (board == null || board.length == 0) return

        val colLength = board.length
        val rowLength = board(0).length
        // 第一遍遍历，将所有边界上的O及与其相连的O设置为#
        for (col <- Array.range(0, colLength)) {
            for (row <- Array.range(0, rowLength)) {
                if ((col == 0 || col == colLength - 1 || row == 0 || row == rowLength - 1)
                    && board(col)(row) == 'O') {
                    dfs(col, row)
                }
            }
        }
        // 第二遍遍历，将未与边界相连的O设置为X，并将#设置回O
        for (col <- Array.range(0, colLength)) {
            for (row <- Array.range(0, rowLength)) {
                if (board(col)(row) == 'O') board(col)(row) = 'X'
                if (board(col)(row) == '#') board(col)(row) = 'O'
            }
        }

        /**
         * 将与边界0相连的所有O设置为# 第二遍扫描时再设置回O
         * @param col The col cur index
         * @param row The row cur Index
         */
        def dfs(col: Int, row: Int): Unit = {
            if (col < 0 || row < 0 || col >= colLength || row >= rowLength
                || board(col)(row) == 'X' || board(col)(row) == '#') {
                // 如果等于# 说明已经搜索过了
                return
            }
            board(col)(row) = '#'
            dfs(col - 1, row)
            dfs(col + 1, row)
            dfs(col, row - 1)
            dfs(col, row + 1)
        }
    }


    def main(args: Array[String]): Unit = {
        val board = Array(Array('O', 'O'), Array('O', 'O'))
        solve(board)
        board.foreach(row => {
            row.foreach(print(_))
            println()
        })
    }

}
